### ISSUE TITLE
Revisit the SCons special case

### ISSUE THREAD

**Issue by RyanGlScott** [goldfirere/singletons#234]
Currently, we special-case the singleton constructor for `(:)` as `SCons`. But we don't actually need to - we could just as well follow the normal singling rules for constructors and single `(:)` to `(:%)`.

Still, `SCons` has had a long history in `singletons`, and it does provide a nice, alphanumeric parallel with its special-case counterpart `SNil`. Perhaps we should still offer `SCons` via a pattern synonym.

**Comment by RyanGlScott** [comment]
I recently implemented this, and removing the special case for `SCons` was as straightforward as one might expect. However, I stumbled upon some difficulties in making the `SCons` pattern synonym work robustly:

1. Due to [GHC Trac #14058](https://ghc.haskell.org/trac/ghc/ticket/14058), one cannot bundle `SCons` with the other `SList` constructors. That is, you can't export `Sing(SNil, (:%), SCons))`, which means that users who write `import Data.Singletons.Prelude (Sing(..))` won't bring `SCons` into scope. This is annoying, but perhaps not a deal-breaker.
2. The larger issue comes from [GHC Trac #14059](https://ghc.haskell.org/trac/ghc/ticket/14059). Essentially, `COMPLETE` sets don't work at all for pattern synonyms for data family instance constructors, so even if we have `{-# COMPLETE SNil, SCons #-}`, GHC won't warn that functions like this are non-exhaustive:

   ```haskell
   f :: Sing (z :: [a]) -> Int
   f SCons{} = 0
   ```

   This makes me extremely nervous, since this means that those who choose to continue using `SCons` won't get the pattern-match warnings that `singletons` programmers so desperately need.

In short, I'm not sure I want to pull the trigger on this change until at least one (or both) of these GHC bugs are fixed. What are your thoughts?

**Comment by goldfirere** [comment]
I wouldn't do this. Perhaps once these bugs are fixed, then it makes sense, but this is an unforced breaking chance with negligible benefit.

**Comment by RyanGlScott** [comment]
I'm forced to agree that this should wait. Having the `SCons` case makes some code uglier in #241, but I can live with it until the aforementioned bugs are fixed.

**Comment by RyanGlScott** [comment]
This would conceivably be unblocked if #318 were implemented, as we'd no longer be limited by `Sing` being a data family. But #318 has its own set of issues to resolve—see https://github.com/goldfirere/singletons/issues/318#issuecomment-375897686.

**Comment by RyanGlScott** [comment]
I've decided that I no longer want this. The complications of using pattern synonyms are just too much.
